### PR TITLE
[framework] URI encode key for asset bundle

### LINKED ISSUE #109195: Image.asset throws `Unable to load asset` when the filename contains space
## Steps to Reproduce

Clone https://github.com/Anas35/Demo and run in latest beta verison, You will see Actual result.

Change version and run the same application in the stable branch, You will Expected result. 

**Expected results:** 

Screenshot_1660017428

**Actual results:** 

Screenshot_1660016779

*Exception:*

```
════════ Exception caught by image resource service ════════════════════════════
The following assertion was thrown resolving an image codec:
Unable to load asset: assets/Flutter Logo.png.

When the exception was thrown, this was the stack
#0      PlatformAssetBundle.loadBuffer
#1      AssetBundleImageProvider._loadAsync
#2      AssetBundleImageProvider.loadBuffer
#3      ImageProvider.resolveStreamForKey.<anonymous closure>
#4      ImageCache.putIfAbsent
#5      ImageProvider.resolveStreamForKey
#6      ScrollAwareImageProvider.resolveStreamForKey
#7      ImageProvider.resolve.<anonymous closure>
#8      ImageProvider._createErrorHandlerAndKey.<anonymous closure>
<asynchronous suspension>
Image provider: AssetImage(bundle: null, name: "assets/Flutter Logo.png")
Image key: AssetBundleImageKey(bundle: PlatformAssetBundle#71b20(), name: "assets/Flutter Logo.png", scale: 1.0)
════════════════════════════════════════════════════════════════════════════════
```
 
**Flutter doctor:**
```
flutter doctor -v
[√] Flutter (Channel beta, 3.3.0-0.2.pre, on Microsoft Windows [Version 10.0.19043.1826], locale en-IN)
    • Flutter version 3.3.0-0.2.pre on channel beta at C:\src\flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 7ac27ac8e6 (6 days ago), 2022-08-02 14:35:08 -0700
    • Engine revision d1e7dc18bf
    • Dart version 2.18.0 (build 2.18.0-271.4.beta)
    • DevTools version 2.15.0

[√] Android toolchain - develop for Android devices (Android SDK version 33.0.0)
    • Android SDK at -\AppData\Local\Android\sdk
    • Platform android-33, build-tools 33.0.0
    • Java binary at: C:\Program Files\Android\Android Studio\jre\bin\java
    • Java version OpenJDK Runtime Environment (build 11.0.12+7-b1504.28-7817840)
    • All Android licenses accepted.

[√] Chrome - develop for the web
    • Chrome at C:\Program Files\Google\Chrome\Application\chrome.exe

[X] Visual Studio - develop for Windows
    X Visual Studio not installed; this is necessary for Windows development.
      Download at https://visualstudio.microsoft.com/downloads/.
      Please install the "Desktop development with C++" workload, including all of its default components

[√] Android Studio (version 2021.2)
    • Android Studio at C:\Program Files\Android\Android Studio
    • Flutter plugin can be installed from:
       https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
       https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.12+7-b1504.28-7817840)

[√] IntelliJ IDEA Community Edition (version 2022.1)
    • IntelliJ at C:\Program Files\JetBrains\IntelliJ IDEA Community Edition 2022.1.4
    • Flutter plugin can be installed from:
       https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
       https://plugins.jetbrains.com/plugin/6351-dart

[√] VS Code (version 1.70.0)
    • VS Code at -\AppData\Local\Programs\Microsoft VS Code
    • Flutter extension version 3.46.0

[√] Connected device (4 available)
    • sdk gphone x86 (mobile) • emulator-5554 • android-x86    • Android 11 (API 30) (emulator)
    • Windows (desktop)       • windows       • windows-x64    • Microsoft Windows [Version 10.0.19043.1826]
    • Chrome (web)            • chrome        • web-javascript • Google Chrome 103.0.5060.134
    • Edge (web)              • edge          • web-javascript • Microsoft Edge 103.0.1264.44

[√] HTTP Host Availability
    • All required HTTP hosts are available
```

### PR DESCRIPTION
**Pull request by jonahwilliams**
Fixes https://github.com/flutter/flutter/issues/109195

I missed the URI encoding applied to the asset key when migrating to ImmutableBuffer.fromAsset

### CODE REVIEW

File: packages/flutter/lib/src/services/asset_bundle.dart [hunk 1]
```diff
@@ -267,7 +267,8 @@ class PlatformAssetBundle extends CachingAssetBundle {
       return ui.ImmutableBuffer.fromUint8List(bytes.buffer.asUint8List());
     }
     try {
-      return await ui.ImmutableBuffer.fromAsset(key);
+      final String encodedKey = Uri(path: Uri.encodeFull(key)).path;
```
Comment (jonahwilliams) [inline packages/flutter/lib/src/services/asset_bundle.dart:270]: Technically we _can_ load an asset with a space in the name just fine - but historically we couldn't, so the tool has always done URI encoding when bundling. To match the current behavior, we need to perform the same encoding here

Comment (jonahwilliams) [reply packages/flutter/lib/src/services/asset_bundle.dart:270]: We could remove the URI encoding from the tool, but that might get weird , and we may have other undiscovered bugs

Comment (zanderso) [reply packages/flutter/lib/src/services/asset_bundle.dart:270]: Is this the only call to `ImmutableBuffer.fromAsset()` that needs this fix? I'm also wondering if it's needed for `FragmentProgram.fromAsset()` which uses the asset key ~exactly the same way that `ImmutableBuffer` does.

**Review by zanderso — APPROVED** [review]
lgtm w/ question

**Comment by jonahwilliams** [comment]
Going to approach this in a different way

**Pull request closed without merging**